src/api/app/models/workflow/step/set_flags.rb

```diff
@@ -0,0 +1,55 @@
+class Workflow::Step::SetFlags < ::Workflow::Step
+  REQUIRED_KEYS = [:flags].freeze
+  REQUIRED_FLAG_KEYS = [:type, :status, :project].freeze
+  OPTIONAL_FLAG_KEYS = [:package, :repository, :architecture].freeze
+
+  validate :validate_project_and_package
+  validate :validate_flags
+
+  def call(_options = {})
+    return unless valid?
+
+    set_flags
+  end
+
+  def flags
+    step_instructions[:flags]
+  end
+
+  private
+
+  def set_flags
+    ActiveRecord::Base.transaction do
+      flags.each do |flag|
+        project = Project.find_by(name: flag[:project])
+        package = project.packages.find_by(name: flag[:package]) if project.present? && flag[:package].present?
+        main_object = package.presence || project
+        check_access(main_object)
+        main_object.add_flag(flag[:type], flag[:status], flag[:repository], flag[:architecture]).save
+      end
+    end
+  end
+
+  def check_access(object)
+    raise Pundit::NotAuthorizedError unless Pundit.policy(token.user, object).update?
+  end
+
+  # rubocop:disable Metrics/CyclomaticComplexity
+  def validate_project_and_package
+    flags.each do |flag|
+      project = Project.find_by(name: flag[:project])
+      package_error = !(project.nil? || flag[:package].nil? || (flag[:package].present? && project.present? && project.packages.find_by(name: flag[:package]).present?))
```

Comment:
This is basically not understandable. How about not lumping those many checks together?

Reply:
Agree with @hennevogel. IMHO, you could create a method, something like:

```ruby
def package_error?(project, flag)
  return false if project.nil? || flag[:package].nil?
  return false if flag[:package].present? && project.present? && project.packages.find_by(name: flag[:package]).present?
  return true
end
```

This should also reduce the `Metrics/CyclomaticComplexity`